### DISCUSSION TITLE
Editable Description in Outline View

### DISCUSSION THREAD

**Discussion by kapishsingla** [eclipse-theia/theia#13188] [Q&A]
Hi, 
I'm working on enhancing Theia's Outline view for GLSP nodes. My goal is to allow users to edit a node's description by clicking on it and opening an editable Outline view. The outline view will show the description of respective node and it will be editable in outline view. So what will be right approach to implement it?

<img width="285" alt="image" src="https://github.com/eclipse-theia/theia/assets/132569352/afbc0999-c71b-4c3f-a70e-5f1c99f3b415">

**Comment by planger** [comment]
Hi @kapishsingla,

Thank you for your question! There is two separate aspects in your question:
1. Bring data from an editor (such as a GLSP editor) into the outline view
2. Allow users to edit nodes in the outline view

Implementing (1) is rather straightforward. For the Theia part, you'd need to inject [`OutlineViewService`](https://github.com/theia-ide/theia/blob/63d66679f5a681a83affb7361333134d5d55f8cd/packages/outline-view/src/browser/outline-view-service.ts) and invoke its method `publish(roots)` with the elements you want to show there. To get the data from the GLSP diagram, one approach would be to register a dedicated service that listens to the active editor, check whether the active editor is your GLSP editor, query its outline (probably by sending a custom GLSP action to the server and wait for the result), and then publish this result to the outline view.
This can be setup somewhat similar to the [problem marker integration](https://github.com/eclipse-glsp/glsp-theia-integration/blob/v2.0.0/packages/theia-integration/src/browser/diagram/theia-marker-manager.ts) in GLSP's Theia integration.

Regarding (2): I don't think the generic outline view in Theia actually supports editing at all. So this would have to be implemented by customizing this view for your purposes. I'm not sure whether this is a great approach though, as it changes a pretty central view in Theia, for a single editor. Depending on your use case, this might not be an issue, but just saying.

Alternatively, you may rather want to populate [Theia's extensible property view](https://theia-ide.org/docs/property_view/) for this purpose or create a custom view from scratch that serves this functionality?
